src/flash/nor/rp2040.c

```diff
@@ -172,6 +172,23 @@ static int rp2040_flash_write(struct flash_bank *bank, const uint8_t *buffer, ui
 		return ERROR_TARGET_UNALIGNED_ACCESS;
 	}
 
+	// This is a good place to do this because flash erase always follows a reset init
+	LOG_DEBUG("Connecting internal flash");
+	err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_CONNECT_INTERNAL_FLASH, NULL, 0);
+	if (err != ERROR_OK)
+	{
+		LOG_ERROR("RP2040 write: failed to connect internal flash");
+		return err;
+	}
```

Comment:
I am not sure if this is really necessary for write

Reply:
Yeah, _probably_ no need to connect flash inside write, as I guess there'll always be an erase (some time) before a write.
Probably also no need to enter XIP mode at the end of erase, because there'll always be a write (some time) after an erase?
:man_shrugging: (hopefully @kilograham will be able to confirm)

Reply:
Entering XPI mode is necessary as the sequence of commands that gdb issues when writing discontinuous flash regions is erase, erase, write, write.

This is why we submitted this PR.

Reply:
I could be wrong, but AIUI the code previously assumed that the sequence would always be: [erase, write, erase, write] and with erase exiting XIP (before doing the erase), and write entering XIP (after doing the write) this would lead to:
```
exit-XIP
do-erase  # flash not in XIP mode

do-write  # flash not in XIP mode
enter-XIP

exit-XIP
do-erase  # flash not in XIP mode

do-write  # flash not in XIP mode
enter-XIP
# flash in XIP mode, execution can continue
```

However as you've discovered the sequence can _actually_ be [erase, erase, write, write] which leads to the erroneous sequence:
```
exit-XIP
do-erase  # flash not in XIP mode

exit-XIP
do-erase  # flash not in XIP mode

do-write  # flash not in XIP mode
enter-XIP

do-write  # flash in XIP mode - ERROR!
enter-XIP
```
where the second do-write fails, because the flash is still in XIP mode.

This PR changes the sequence to:
```
exit-XIP
do-erase  # flash not in XIP mode
enter-XIP

exit-XIP
do-erase  # flash not in XIP mode
enter-XIP

exit-XIP
do-write  # flash not in XIP mode
enter-XIP

exit-XIP
do-write  # flash not in XIP mode
enter-XIP
# flash in XIP mode, execution can continue
```

What I was suggesting with my comment above was that it's probably not necessary for the erase command to enter XIP mode, which would then lead to this sequence of commands:
```
exit-XIP
do-erase  # flash not in XIP mode

exit-XIP
do-erase  # flash not in XIP mode

exit-XIP
do-write  # flash not in XIP mode
enter-XIP

exit-XIP
do-write  # flash not in XIP mode
enter-XIP
# flash in XIP mode, execution can continue
```

I hope that makes sense? (of course this assumes that you'd never get a [erase, erase, write, write, erase] sequence _without_ a write at the very end)

Reply:
If that works, yes. I'm not sure it works though, but I will test and get back to you.

Reply:
> I hope that makes sense? (of course this assumes that you'd never get a [erase, erase, write, write, erase] sequence _without_ a write at the very end)

The point is you *don't want* to be making assumptions about the order of calls over which you have no control

Reply:
> The point is you don't want to be making assumptions about the order of calls over which you have no control

Fair enough - perhaps that means that we _do_ also want to connect-flash from inside the write function then?

Maybe it's worth factoring out a couple of small functions (one to connect-flash and exit-xip, and another to enter-xip) which could be called from both the write function and the erase function? :shrug:

Reply:
Something like:

```c
static int rp2040_flash_init(struct flash_bank *bank)
{
        struct rp2040_flash_bank *priv = bank->driver_priv;
        err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_CONNECT_INTERNAL_FLASH, NULL, 0);
	if (err != ERROR_OK)
	{
		LOG_ERROR("RP2040 write: failed to connect internal flash");
		return err;
	}
        return err;
}

static int rp2040_flash_exit_xip(struct flash_bank *bank)
{
        struct rp2040_flash_bank *priv = bank->driver_priv;
        err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_FLASH_EXIT_XIP, NULL, 0);
	if (err != ERROR_OK)
	{
		LOG_ERROR("RP2040 write: failed to exit flash XIP mode");
		return err;
	}
        return err;
}

static int rp2040_flash_enter_exip(struct flash_bank *bank)
{
        struct rp2040_flash_bank *priv = bank->driver_priv;
        err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_FLASH_ENTER_CMD_XIP, NULL, 0);
	if (err != ERROR_OK)
	{
		LOG_ERROR("RP2040 write: failed enter flash XIP mode");
	}
        return err;
}
```

Reply:
Yeah but I was thinking you could combine your `rp2040_flash_init` and `rp2040_flash_exit_xip` functions into a single function, since they always get used together? :shrug:

Reply:
If this is OK with everyone, I'll add it to the code.

```c
static int rp2040_flash_exit_xip(struct flash_bank *bank)
{
        struct rp2040_flash_bank *priv = bank->driver_priv;
        err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_CONNECT_INTERNAL_FLASH, NULL, 0);
	if (err != ERROR_OK)
	{
		LOG_ERROR("RP2040 write: failed to connect internal flash");
		return err;
	}
        err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_FLASH_EXIT_XIP, NULL, 0);
	if (err != ERROR_OK)
	{
		LOG_ERROR("RP2040 write: failed to exit flash XIP mode");
		return err;
	}
        return err;
}

static int rp2040_flash_enter_exip(struct flash_bank *bank)
{
        struct rp2040_flash_bank *priv = bank->driver_priv;
        err = rp2040_call_rom_func(bank->target, priv->stacktop, FUNC_FLASH_ENTER_CMD_XIP, NULL, 0);
	if (err != ERROR_OK)
	{
		LOG_ERROR("RP2040 write: failed enter flash XIP mode");
	}
        return err;
}
```

Reply:
a) you probably want to change the error-strings to say something like "RP2040 exit_xip: ..." and "RP2040 enter_xip: ..." ? (and perhaps change "failed enter" to "failed to enter")
b) I think you want to add `int err = ERROR_OK;` near the top of each of those functions
c) `rp2040_flash_enter_exip` -> `rp2040_flash_enter_xip` ? :wink: 
d) it might be worth waiting for the go-ahead from @kilograham

Reply:
> a) you probably want to change the error-strings to say something like "RP2040 exit_xip: ..." and "RP2040 enter_xip: ..." ? (and perhaps change "failed enter" to "failed to enter")
Sure :)
> b) I think you want to add `int err = ERROR_OK;` near the top of each of those functions
I think we can add int `err = rp2040_call_rom_func ...`
> c) `rp2040_flash_enter_exip` -> `rp2040_flash_enter_xip` ? 😉
:)
> d) it might be worth waiting for the go-ahead from @kilograham

Reply:
Dome